### DISCUSSION TITLE
How do I make a generic that takes a `TSchema` or `undefined` and if undefined returns `TUnknown` while still retaining type inference?

### DISCUSSION THREAD

**Discussion by jessekrubin** [sinclairzx81/typebox#424] [Q&A]
I am trying to write a generic function that takes a `TSchema` param `const makeSchema = <T extends TSchema>(schema?: T)` but returns `TUnknown` if the schema given is undefined. Maybe this is a dumb question, but I am confused!

**Comment by sinclairzx81** [comment]
@jessekrubin Hi!

Unfortunately, there's not really an ideal way to define optional arguments for generic types. The problem with optionals is that `schema?` could be either `TSchema` or `undefined`. Because of this, the body of `makeSchema()` will need (at a minimum) at least two `return` code paths, one to handle the `undefined` case, the other to handle the `TSchema` case. As there are multiple return types possible, things can get somewhat complicated.

The following implements some conditional types to map for the `undefined` / optional case. I'm not sure what the schema should look like in the instance it's `undefined` but have added the `TObject<{}>` as a placeholder for you to change with the actual type. Note the similarities between the conditional type and the body of the `Make` function.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBoUGUDGAFhiAIY7IDyARgFYb4w4CCAzqxrGpnAL5wAzKBBBwA5AAFWwAHb4ANqWBQA9DHQZqEAB5iAsAChD6ngElWAEQwDZGACYAeZAD44AXjgBRbTCilGTjgArjJ21rZ2rhg+GKGscDIYAG6ccAD8cL5BGHAAXIKk8hzGGigAsqQA1hhOcNEwsXbxyATEZHAAPnAhYTaJke5w5lZ99k5RMXGZUNlwhnAL6Sg09AFI2vlsHFwa43yu84v5VHQMMA4IvM6GhvgQMqzwFdWDtfWNza0kpJ3doeH9ZwAClYRG+aWOAEp3K4EIcoBgYEEoDI4CCwe0AIRuDw9AH2OYGRZLbgYAB0KzOQPW+VBbR+vGhhwWxw0FNOjGpfEhh2hpGazxqLkMvBuBjuD3gzEGgqBpLJuF8sgA5kDIdC4CoVMsOecaXglTJlQBufa3e6POAAIRlVQwauJjqdWp1qz1ZqMnqAA)

```typescript
import { Type, TSchema, TObject, AssertType } from '@sinclair/typebox'

type IsDefined<T> = Extract<T, undefined> extends never ? true : false
type TMake<T extends TSchema | undefined> = IsDefined<T> extends true 
    ? TObject<{ x: AssertType<T> }> 
    : TObject<{}>

const Make = <T extends TSchema | undefined>(schema?: T) => {
  return (schema !== undefined 
    ? Type.Object({ x: schema }) 
    : Type.Object({ })
  ) as TMake<T>
}

const A = Make(Type.String())  // TObject<{ x: TString; }>
const B = Make()               // TObject<{ }>
```
Hopefully you can update the above for your use case!
Cheers
S

**Comment by jessekrubin** [comment]
@sinclairzx81 I just made public the package/code I have been fiddling with in my spare time (https://github.com/jessekrubin/geobox) and I am trying to figure out an elegant way of being able to hand the feature-schema-creating-function (eg: `PointFeature`) the coordinate schema (which is either 2d or 3d) while still retaining the ts-type hints,

Ideally if no schema is provided as an option to the function (`PointFeature()` vs `PointFeature(Type.Tuple([Type.Number(), Type.Number()]))`) then the schema falls back to the union of 2d/3d schemas. 

This is in no way pressing, just curious if you have any thoughts!

**Reply by sinclairzx81** [reply]
@jessekrubin Hi, your project looks cool! Nice work! :)

> Ideally if no schema is provided as an option to the function (PointFeature() vs PointFeature(Type.Tuple([Type.Number(), Type.Number()]))) then the schema falls back to the union of 2d/3d schemas.

There's a few ways to approach this, but generally the approach I take is to use overloaded functions + explicit return. Also, there's a few ways to approach generic type representation for `PointFeature`. Below is a couple of options.

### Option A: Overloads + Instantiation Expressions

This approach uses function overloads and defines a template type for generic type `PointFeature<T>`. It uses [TS 4.7 Instantiation Expressions](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#instantiation-expressions) to derive the TB types required for inference.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBoUGUDGAFhiAIY64ykzD5wC+cAZlBCHAOQACAzsAHb4ANqWBQA9DHQYARhAAeHALAAoVePFwAtDt179Bw0eMnj6zQGEI0ACYDqGHue2nXb929UZ5kWHHwQ-DzwAAoQAjAAzAAicAC8KNIAdMgArmBCGAAUANpomEkAcqkgMhhQWQCUOPkYRSVlFdWJBcWl5VUAupVePtDwAUGh4fwwAEyxCbUp6Zm5022NVTXJix2V3b2+A4HBcGER8S11AKr8wIG5B6MTONdR0ZtqKhouHu8f7s61KCQZDnAsgBJIakUbAagXfhwACiPigjj4gR6L00n3RGMMW36cCkmBQ9wAYhhqKkEQAeZBwbwwDD8Gw8PBEEikAB8RwAShgYGT+LVyXiMBAmASRjBiaSKchWazsX4mKlBDRAqKIhKeVLqfJafTGcgCMQyEd9cyyKysjxTaQAFwoSqIVRwOAIjXQ6YAeRkACsMPgYFkEI6nbjpLbpgAZYC0qCkIRZDjqskYDjVINO4A2MPJM5Q+bJShQAQAc2Wx3q7Sa3SwabgYFYmFgwEctsthtIQfoKPoqmcmL7mOc7oAbuUhBBSAze-3p+8eyoFUqofsxYmEVUw0SSRqMALpMLlxFZfPFX6l5vJTuqTS6QymW3zXWIA2aM27RuV1uk5SjwvTyrz9ulJajqt4mveWSPs+TY8AA-GG9qBioTouryqqjKu2SQeUL6MgAPrhB6jF2c6vDOZFmKisLyKQ4CZFO5EMboc6DHsVIJABSZVKogooEclCQvgu6YPu0rMbs8AAEJHBxa4et6vr+kgdj8EWEBZgUBbFlUDCVCiPFSQk-E0IJgr7hJrJAA)

```typescript
import { Type, TSchema, Static } from '@sinclair/typebox'

// ----------------------------------------------------------------
// Coordinates
// ----------------------------------------------------------------
export const Point3D = Type.Tuple([Type.Number(), Type.Number(), Type.Number()])
export const Point2D = Type.Tuple([Type.Number(), Type.Number()])
export const Point = Type.Union([Point2D, Point3D])

// ----------------------------------------------------------------
// Type Template (Instantiation Expression)
// ----------------------------------------------------------------
export type TPointFeature<T extends TSchema> = ReturnType<typeof TPointFeature<T>> // derived: instantiation expression
export function TPointFeature<T extends TSchema = TSchema>(schema: T) {
  return Type.Object({
    type: Type.Literal('Feature'),
    id: Type.Union([Type.String(), Type.Number()]),
    properties: schema
  })
}

// ----------------------------------------------------------------
// Overloads
// ----------------------------------------------------------------

function PointFeature(): TPointFeature<typeof Point>
function PointFeature<T extends TSchema>(properties: T): TPointFeature<T>
function PointFeature<T extends TSchema>(properties?: T) {
  return TPointFeature(properties || Point)
}

// ------------------------------------------------------------
// Example
// ------------------------------------------------------------

const T = PointFeature()
type T = Static<typeof T>

const B = PointFeature(Type.Object({ dingo: Type.String() }))
type B = Static<typeof B>
```
### Option B: Overloads + Explicit Generic Types

If you're working in older versions of the compiler (or concerned about backwards compatibility to version 4), the following does the same but explicitly specifies the TB types. This should work in all versions of TypeScript 4.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBoUGUDGAFhiAIY64ykzD47IDyARgFYb4z0CqAdsBD3oAZYDAxRSAG3qUowHgHN6AOQCuIJuLgBfOADMoEEHADkAAQDO8-JNLAoAehjoMTCAA8TAWABQvhw5wALQhoWHhEZFR0THR-oEAwhDQACby1BgW8cGxuXn5eb4Y7pCwcPgCFvAAChDyMADMACJwALwoLgB0yKpgkhgAFADaaJidahriAwCU9F0TmlAzc2MLU9MAutNFJdDwFTxVcLX1AEwt7aMY3b39w1fj6ovLHatP61s7pfuVNXU88EuXV4-B4wxOAPOOAhjSanz8PgCOQKKNRaNR2QAoiVJLR4Fc4AMAEoYACOqnsmTgoG+8gUcBgxBQACEGS4LHBSDwUnAoBgwHyLBgAXSGUznJg4MU+rRRJJUNtEYF0SrVWrQl89mzJcgYQAxDDUVR8gA8yCl7jE3I5yAIxDIAD42ihmGwOCaEL44NqMAAuFAiMQSSQmkwGo18kwOr3UlL+5AggQmkayOkqd5QDbRnzegUQTCwYCZeO+bTZ7LqysYpVwBgAN3EkggpBSWRrVY7hQRelUPA4oOO-xg4ZgxsG03j+sNo9NEowED0g-q2Z7fZoAiXAJHY7NFqtrbwRBIpAdAzzBZoxZQE5QU4jGDNK97-Y3d5nD-NxX3Nrtx9P5-ES8LAAfnjaZEBjPkZx4V5rlddgYAGT0c29H14y6QNxCkAYw2nMcTFmGNvWAONYM6RMwRGLpU0UF4HjWJZNkIlDc0MC8iwsf0AMLKkAB9eM3GAY20RVtF8CtO0ksIsXcUhwH6CSpKk8SfAOI5zXaN8xxmXw5xQZ1KGoWgTTnBcUHLVTfjgVlNKHbc+QGB54I4JC4DSRQIHQsYaIUGYdGmRU9JsuBDJofATJcMzmQdIA)

```typescript
import { Type, TSchema, Static, TObject, TUnion, TLiteral, TString, TNumber } from '@sinclair/typebox'

// ----------------------------------------------------------------
// Coordinates
// ----------------------------------------------------------------
export const Point3D = Type.Tuple([Type.Number(), Type.Number(), Type.Number()])
export const Point2D = Type.Tuple([Type.Number(), Type.Number()])
export const Point = Type.Union([Point2D, Point3D])

// -------------------------------------------------------------------------------------
// Explict Type (Requires importing the TB types and representing the type explicitly)
// -------------------------------------------------------------------------------------
export type TPointFeature<T extends TSchema> = TObject<{  // explicit
  type: TLiteral<'Feature'>
  id: TUnion<[TString, TNumber]>
  properties: T
}>

// ----------------------------------------------------------------
// Overloads
// ----------------------------------------------------------------

function PointFeature(): TPointFeature<typeof Point>
function PointFeature<T extends TSchema>(properties: T): TPointFeature<T>
function PointFeature<T extends TSchema>(properties?: T) {
  return Type.Object({
    type: Type.Literal('Feature'),
    id: Type.Union([Type.String(), Type.Number()]),
    properties: properties || Point
  })
}

// ------------------------------------------------------------
// Example
// ------------------------------------------------------------

const T = PointFeature()
type T = Static<typeof T>

const B = PointFeature(Type.Object({ dingo: Type.String() }))
type B = Static<typeof B>
```

Hope this helps!
S

**Reply by jessekrubin** [reply]
Hi @sinclairzx81

I updated my lib with your suggestions and it works awesome! Making geojson schemas works quite well! All the functions now take an optional "schemas" param which lets you define some or all of the properties/coordinate/bbox schemas and creates your schema(s) with correct type inference! 

Would love if you would glance over it and for your feedback if you have the time. Wonder what types of style-things you would change to make it more "typebox"-y

Also wonder what recommendations you might have for making it Deno compatible. 

Also, I started on a alternative version using zod (called "geozod") but that is proving more difficult than I thought it would be. 

Regards,
Jesse

PS: do you have any benchmarks of value-check against zod?

**Reply by sinclairzx81** [reply]
@jessekrubin Hey, nice work :) Ill try take a look over the weekend 

> PS: do you have any benchmarks of value-check against zod?

I don't have any value check benchmarks comparing Zod, but do keep track of compiler benchmarks here https://sinclairzx81.github.io/runtime-type-benchmarks/. Could look into adding the value checkers to the suite, but probably not for a while (currently on other projects for the next few weeks and have been fairly flat out with other things recently), will make a note to include them when I next get some time.

Good work :)

**Reply by jessekrubin** [reply]
Just curious if you have a suggestion for how to make a field required depending on another field,

In the case of a `tilejson` (https://github.com/mapbox/tilejson-spec/blob/master/3.0.0/schema.json), when `format: 'png'|'jpg'`, the `vector_layers` key is optional, but when `format: 'pbf'`, `vector_layers` is required.

The official json schema has `vector_layers` as a required key, but the readme affiliated with the spec says `vector_layers` is only required when format is `pbf`

**Reply by sinclairzx81** [reply]
> Just curious if you have a suggestion for how to make a field required depending on another field,

Currently, the TypeBox validators don't have support for conditional `if/then/else` types (or `dependentRequired`). However you can still express these schematics in TypeBox, but will need to use Ajv to validate them.

```typescript
import { Type } from '@sinclair/typebox'

const T = Type.Object({
  format: Type.Union([
    Type.Literal('png'),
    Type.Literal('jpg'),
    Type.Literal('pdf')
  ])
}, {
  if: Type.Object({
    format: Type.Literal('pdf')             // if format === 'pdf'
  }),
  then: Type.Object({
    vector_layers: Type.Array(Type.Any())   // then expect this property
  })
})

import Ajv from 'ajv'

const ajv = new Ajv()
console.log(ajv.validate(T, { format: 'png' }))                           // true
console.log(ajv.validate(T, { format: 'jpg' }))                           // true
console.log(ajv.validate(T, { format: 'pdf' }))                           // false
console.log(ajv.validate(T, { format: 'pdf', vector_layers: [1, 2, 3] })) // true
```
Conditional schematics are a planned feature for TB, but probably won't be ready for some time.
Hope this helps!
S

**Reply by jessekrubin** [reply]
Thanks for your (speedy as always) reply! Would the `TypeCompiler` be able to compile/validate the following:

```typescript

export const VectorLayer = Type.Object(
  {
    id: Type.String(),
    fields: Type.Record(Type.String(), Type.Any()),
    description: Type.Optional(Type.Union([Type.String(), Type.Null()])),
    maxzoom: Type.Optional(Type.Union([Zoom, Type.Null()])),
    minzoom: Type.Optional(Type.Union([Zoom, Type.Null()])),
  },
  {
    additionalProperties: true,
  }
);

export const VectorLayers = Type.Array(VectorLayer, {
  default: [],
});

export const Tilejson300Base = Type.Object({
  // MUST
  name: Type.String(),
  tilejson: Type.String({ default: '3.0.0' }),
  tiles: Type.Array(Type.String(), {
    default: [],
  }),
  // OPTIONAL
  vector_layers: Type.Optional(
    VectorLayers
  ),
  version: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  description: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  minzoom: Type.Optional(Type.Union([Type.Number(), Type.Null()])),
  maxzoom: Type.Optional(Type.Union([Type.Number(), Type.Null()])),
  bounds: Type.Optional(Type.Union([Type.Array(Type.Number()), Type.Null()])),
  center: Type.Optional(Type.Union([Type.Array(Type.Number()), Type.Null()])),
  attribution: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  template: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  scheme: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  legend: Type.Optional(Type.Union([Type.String(), Type.Null()])),
  grids: Type.Optional(Type.Union([Type.Array(Type.String()), Type.Null()])),
  data: Type.Optional(Type.Union([Type.Array(Type.String()), Type.Null()])),
  fillzoom: Type.Optional(Type.Union([Zoom, Type.Null()])),
});
export const Tilejson300Raster = Type.Composite([
  Type.Object({
    // MUST
    format: Type.Union([Type.Literal('png'), Type.Literal('jpg')]),
  }),
  Tilejson300Base,
]);
export const Tilejson300Vector = Type.Composite([
  Type.Object({
    format: Type.Literal('pbf'),
    vector_layers: VectorLayers,
  }),
  Tilejson300Base,
]);
export const Tilejson300 = Type.Union([Tilejson300Raster, Tilejson300Vector]);
```

**Reply by sinclairzx81** [reply]
@jessekrubin Seems to compile ok. Generally that means it's going to validate as you expect.

```typescript
console.log(TypeCompiler.Code(Tilejson300))

// const local_0 = new RegExp(/^(.*)$/)
// const local_1 = new RegExp(/^(.*)$/)
// const local_2 = new RegExp(/^(.*)$/)
// return function check(value) {
//   return (
//     (((typeof value === 'object' && value !== null && !Array.isArray(value)) && (((value.format === 'png')) || ((value.format === 'jpg'))) && (typeof value.name === 'string') && (typeof value.tilejson === 'string') && Array.isArray(value.tiles) && value.tiles.every((value) => ((typeof value === 'string'))) && (value.vector_layers !== undefined 
// ? (Array.isArray(value.vector_layers) && value.vector_layers.every((value) => ((typeof value === 'object' && value !== null && !Array.isArray(value)) && (typeof value.id === 'string') && (typeof value.fields === 'object' && value.fields !== null && !Array.isArray(value.fields) && !(value.fields instanceof Date) && !(value.fields instanceof Uint8Array)) && (Object.entries(value.fields).every(([key, value]) => (local_0.test(key) ? (true) : true))) && (value.description !== undefined ? ((((typeof value.description === 'string')) || ((value.description === null)))) : true) && (value.maxzoom !== undefined ? ((((typeof value.maxzoom === 'number' && Number.isFinite(value.maxzoom))) || ((value.maxzoom === null)))) : true) && (value.minzoom !== undefined ? ((((typeof value.minzoom === 'number' && Number.isFinite(value.minzoom))) || ((value.minzoom === null)))) : true)))) : true) && (value.version !== undefined ? ((((typeof value.version === 'string')) || ((value.version === null)))) : true) && (value.description !== undefined ? ((((typeof value.description === 'string')) || ((value.description === null)))) : true) && (value.minzoom !== undefined ? ((((typeof value.minzoom === 'number' && Number.isFinite(value.minzoom))) || ((value.minzoom === null)))) : true) && (value.maxzoom !== undefined ? ((((typeof value.maxzoom === 'number' && Number.isFinite(value.maxzoom))) || ((value.maxzoom === null)))) : true) && (value.bounds !== undefined ? (((Array.isArray(value.bounds) && value.bounds.every((value) => ((typeof value === 'number' && Number.isFinite(value))))) || ((value.bounds === null)))) : true) && (value.center !== undefined ? (((Array.isArray(value.center) && value.center.every((value) => ((typeof value === 'number' && Number.isFinite(value))))) || ((value.center === null)))) : true) && (value.attribution !== undefined ? ((((typeof value.attribution === 'string')) || ((value.attribution === null)))) : true) && (value.template !== undefined ? ((((typeof value.template === 'string')) || ((value.template === null)))) : true) && (value.scheme !== undefined ? ((((typeof value.scheme === 'string')) || ((value.scheme === null)))) : true) && (value.legend !== undefined ? ((((typeof value.legend === 'string')) || ((value.legend === null)))) : true) && (value.grids !== undefined ? (((Array.isArray(value.grids) && value.grids.every((value) => ((typeof value === 'string')))) || ((value.grids === null)))) : true) && (value.data !== undefined ? (((Array.isArray(value.data) && value.data.every((value) => ((typeof value === 'string')))) || ((value.data === null)))) : true) && (value.fillzoom !== undefined ? ((((typeof value.fillzoom === 'number' && Number.isFinite(value.fillzoom))) || ((value.fillzoom === null)))) : true)) || ((typeof value === 'object' && value !== null && !Array.isArray(value)) && (value.format === 'pbf') && ((Array.isArray(value.vector_layers) && value.vector_layers.every((value) => ((typeof 
// value === 'object' && value !== null && !Array.isArray(value)) && (typeof value.id === 'string') && (typeof value.fields === 'object' && value.fields !== null && !Array.isArray(value.fields) && !(value.fields instanceof Date) && !(value.fields instanceof Uint8Array)) && (Object.entries(value.fields).every(([key, value]) => (local_1.test(key) ? 
// (true) : true))) && (value.description !== undefined ? ((((typeof value.description === 'string')) || ((value.description === null)))) : true) && (value.maxzoom !== undefined ? ((((typeof value.maxzoom === 'number' && Number.isFinite(value.maxzoom))) || ((value.maxzoom === null)))) : true) && (value.minzoom !== undefined ? ((((typeof value.minzoom === 'number' && Number.isFinite(value.minzoom))) || ((value.minzoom === null)))) : true)))) && (Array.isArray(value.vector_layers) && value.vector_layers.every((value) => ((typeof value === 'object' && value !== null && !Array.isArray(value)) && (typeof value.id === 'string') && (typeof value.fields === 'object' && value.fields !== null && 
// !Array.isArray(value.fields) && !(value.fields instanceof Date) && !(value.fields instanceof Uint8Array)) && (Object.entries(value.fields).every(([key, value]) => (local_2.test(key) ? (true) : true))) && (value.description !== undefined ? ((((typeof value.description === 'string')) || ((value.description === null)))) : true) && (value.maxzoom !== undefined ? ((((typeof value.maxzoom === 'number' && Number.isFinite(value.maxzoom))) || ((value.maxzoom === null)))) : true) && (value.minzoom !== undefined ? ((((typeof value.minzoom === 'number' && Number.isFinite(value.minzoom))) || ((value.minzoom === null)))) : true))))) && (typeof value.name === 'string') && (typeof value.tilejson === 'string') && Array.isArray(value.tiles) && value.tiles.every((value) => ((typeof value === 'string'))) && (value.version !== undefined ? ((((typeof value.version === 'string')) || ((value.version === null)))) : true) && (value.description !== undefined ? ((((typeof value.description === 'string')) || ((value.description === null)))) : true) && (value.minzoom !== undefined ? ((((typeof value.minzoom === 'number' && Number.isFinite(value.minzoom))) || ((value.minzoom === null)))) : true) && (value.maxzoom !== undefined ? ((((typeof value.maxzoom === 'number' && Number.isFinite(value.maxzoom))) || ((value.maxzoom === null)))) : true) && (value.bounds !== undefined ? (((Array.isArray(value.bounds) && value.bounds.every((value) => ((typeof value === 'number' && Number.isFinite(value))))) || ((value.bounds === null)))) : true) && (value.center !== undefined 
// ? (((Array.isArray(value.center) && value.center.every((value) => ((typeof value === 'number' && Number.isFinite(value))))) || ((value.center === null)))) : true) && (value.attribution !== undefined ? ((((typeof value.attribution === 'string')) || ((value.attribution === null)))) : true) && (value.template !== undefined ? ((((typeof value.template === 'string')) || ((value.template === null)))) : true) && (value.scheme !== undefined ? ((((typeof value.scheme === 'string')) || ((value.scheme === null)))) : true) && (value.legend !== undefined ? ((((typeof value.legend === 'string')) || ((value.legend === null)))) : true) && (value.grids !== undefined ? (((Array.isArray(value.grids) && value.grids.every((value) => ((typeof value === 'string')))) || ((value.grids === null)))) : true) && (value.data !== undefined ? (((Array.isArray(value.data) && value.data.every((value) => ((typeof value === 'string')))) || ((value.data === null)))) : true) && (value.fillzoom !== undefined ? ((((typeof value.fillzoom === 'number' && Number.isFinite(value.fillzoom))) || ((value.fillzoom === null)))) : true)))
//  )
// }
```
You should still write minimal tests around these structures if you're unsure of the validation behavior, but if you come across any issues, be sure to submit an GH issue (I will look into it).
Cheers!

**Reply by jessekrubin** [reply]
Sounds like a plan! I think typebox's type system is wildly powerful!
